File: python/subarray_sum_equals_k.py
class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        result: int = 0
        prefix_map: dict[int, int] = {0: 1}  # prefix sum -> freq
        accu: int = 0
        for num in nums:
            accu += num
            remain: int = accu - k
            if remain in prefix_map:
                result += prefix_map[remain]
            if accu not in prefix_map:
                prefix_map[accu] = 1
            else:
                prefix_map[accu] += 1

        return result


def test_subarraySum_case_1():
    # arrange
    nums: list[int] = [1, 1, 1]
    k: int = 2
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.subarraySum(nums, k)

    # assert
    assert expected == actual


def test_subarraySum_case_2():
    # arrange
    nums: list[int] = [1, 2, 3]
    k: int = 3
    expected: int = 2

    # act
    solution = Solution()
    actual = solution.subarraySum(nums, k)

    # assert
    assert expected == actual

File: rust/src/code/subarray_sum_equals_k.rs
struct Solution {}

// [1, 1, 1]
impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        use std::collections::HashMap;
        let mut pre_accu_map: HashMap<i32, i32> = HashMap::new();
        pre_accu_map.insert(0, 1);
        let mut result: i32 = 0;
        let mut accu: i32 = 0;
        for i in 0..nums.len() {
            accu += nums[i];
            let remain = accu - k;
            if let Some(n) = pre_accu_map.get(&remain) {
                result += n;
            }

            pre_accu_map
                .entry(accu)
                .and_modify(|x| {
                    *x += 1;
                })
                .or_insert(1);
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn subarray_sum_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 1, 1];
        let k: i32 = 2;
        let expected: i32 = 2;

        // act
        let actual = Solution::subarray_sum(nums, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn subarray_sum_case_2() {
        // arrange
        let nums: Vec<i32> = vec![1, 2, 3];
        let k: i32 = 3;
        let expected: i32 = 2;

        // act
        let actual = Solution::subarray_sum(nums, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn subarray_sum_case_3() {
        // arrange
        let nums: Vec<i32> = vec![3, 2, 1, 3];
        let k: i32 = 6;
        let expected: i32 = 2;

        // act
        let actual = Solution::subarray_sum(nums, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn subarray_sum_case_4() {
        // arrange
        let nums: Vec<i32> = vec![-1, -1, 1];
        let k: i32 = 0;
        let expected: i32 = 1;

        // act
        let actual = Solution::subarray_sum(nums, k);

        // assert
        assert_eq!(expected, actual);
    }
}
